src/main/java/de/chojo/jdautil/pagination/PageService.java

```diff
@@ -0,0 +1,230 @@
+/*
+ *     SPDX-License-Identifier: AGPL-3.0-only
+ *
+ *     Copyright (C) 2022 RainbowDashLabs and Contributor
+ */
+
+package de.chojo.jdautil.pagination;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import de.chojo.jdautil.localization.ILocalizer;
+import de.chojo.jdautil.pagination.bag.IPageBag;
+import de.chojo.jdautil.pagination.exceptions.EmptyPageBagException;
+import net.dv8tion.jda.api.entities.ChannelType;
+import net.dv8tion.jda.api.entities.Emoji;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.GuildMessageChannel;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.entities.MessageChannel;
+import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
+import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+import net.dv8tion.jda.api.interactions.InteractionHook;
+import net.dv8tion.jda.api.interactions.components.ActionRow;
+import net.dv8tion.jda.api.interactions.components.buttons.Button;
+import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle;
+import net.dv8tion.jda.api.sharding.ShardManager;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class PageService extends ListenerAdapter {
+    private final String previousLabel;
+    private final String nextLabel;
+    private final String previousText;
+    private final String nextText;
+    private final Cache<Long, IPageBag> cache;
+    private final ILocalizer localizer;
+
+    public PageService(String previousLabel, String nextLabel, String previousText, String nextText, Cache<Long, IPageBag> cache, ILocalizer localizer) {
+        this.previousLabel = previousLabel;
+        this.nextLabel = nextLabel;
+        this.previousText = previousText;
+        this.nextText = nextText;
+        this.cache = cache;
+        this.localizer = localizer;
+    }
+
+    /**
+     * Registers a new page on a slash command event. The page will be send as a reply to this event.
+     *
+     * @param event event
+     * @param page  page
+     */
+    public void registerPage(SlashCommandInteractionEvent event, IPageBag page) {
```

Comment:
I would suggest using the `IReplyCallback` interface here, so we can use it for every interaction which supports replies
```suggestion
    public void registerPage(IReplyCallback event, IPageBag page) {
```